### PR TITLE
Random numbers on gpu

### PR DESCRIPTION
**Pull request by havahol**
***This PR should not be merge yet, but PR is created to allow for review***

Contains:
- generation of random numbers (uniform and normal distributed) on GPU
- Generation of perturbed eta field using a SOAR covariance function
- Generation of perturbed hu and hv fields which makes the perturbed ocean state in geostrophic balance given the perturbed eta field.
- Tests for everything by comparing with equivalent CPU code

Missing before PR can be merged:
- Solve a newly discovered problem with ghost cells
- Use the OceanStateNoise class along with a simulator
- Make it possible to merge automatically

### CODE REVIEW

**Comment by havahol** [comment]
Fixed:
- Solve a newly discovered problem with ghost cells
- Use the OceanStateNoise class along with a simulator

Still missing:
- Make it possible to merge automatically

**Comment by havahol** [comment]
This PR is now ready to be merged

File: sim/src/kernels/ocean_noise.opencl [hunk 1]
```diff
@@ -0,0 +1,292 @@
...
+
+    //Index of start of block within domain
+    const int bx = get_local_size(0) * get_group_id(0);
+    const int by = get_local_size(1) * get_group_id(1);
+
+    //Index of cell within domain
+    const int ti = get_global_id(0);
+    const int tj = get_global_id(1);
+
+    const int cutoff = 2;
```
Comment (setmar) [inline sim/src/kernels/ocean_noise.opencl]: Should be an argument to the kernel (even though we most likely will not change it).

File: python/SWESimulators/OceanStateNoise.py [hunk 2]
```diff
@@ -0,0 +1,409 @@
...
+        self.seed = None
+        
+        self.cl_ctx = cl_ctx
+        self.cl_queue = cl_queue
+        
+        self.nx = np.int32(nx)
+        self.ny = np.int32(ny)
+        self.dx = np.float32(dx)
+        self.dy = np.float32(dy)
+        self.staggered = np.int(0)
```
[the remark this answers is not in the document]
Comment (havahol) [reply python/SWESimulators/OceanStateNoise.py]: To have better control of the data type that goes into the opencl kernel

File: python/SWESimulators/OceanStateNoise.py [hunk 3]
```diff
@@ -0,0 +1,409 @@
...
+        if staggered:
+            self.staggered = np.int(1)
+            
+        # The cutoff parameter is hard-coded.
+        # The size of the cutoff determines the computational radius in the
+        # SOAR function. Hence, the size of the local memory in the OpenCL 
+        # kernels has to be hard-coded.
+        self.cutoff = np.int32(2)
```
[the remark this answers is not in the document]
Comment (havahol) [reply python/SWESimulators/OceanStateNoise.py]: Using this method:
https://docs.python.org/2/faq/programming.html#how-do-i-share-global-variables-across-modules

File: sim/src/kernels/ocean_noise.opencl [hunk 4]
```diff
@@ -0,0 +1,292 @@
...
+    return (float2)(n1, n2);
+}
+
+/**
+  * Kernel that generates uniform random numbers.
+  */
+__kernel void uniformDistribution(
+        // Size of data
+        int seed_nx_, int seed_ny_,
+	int random_nx_,
```
[the remark this answers is not in the document]
Comment (havahol) [reply sim/src/kernels/ocean_noise.opencl]: Did the emacs magic to untabify the given file

**Comment by havahol** [comment]
All comments followed up on, @setmar

**Pull request merged**